### PR TITLE
PMP: Self-intersection local removal

### PR DESCRIPTION
**Pull request by MaelRL**
## Summary of Changes

This PR does the following:
- splits the too eclectic and humongous file `repair.h` into three files: `manifoldness.h`, `remove_degeneracies.h` (contains degenerate edges/faces collectors and removal tools), and `remove_self_intersections.h`. `repair.h` stays where it is (still contains some basic repair functions), and includes the new headers for backward compatibility.
- Add a face range parameter to self intersection
- Add a Boolean such that intersections are only resolved within their CC
- Some PMP::smooth_mesh improvements

TODO:
- [x] implement hole filling / smoothing with constraints
- [x] finish the code (how to determine automatically the Hausdorff bound, pass the local Boolean as parameter, etc.)
- [x] Re-add a function that does global self-intersection removal
- [x] test the code more thoroughly
- [x] clean example/test (some uncommitted data files)

## Release Management

* Affected package(s): `Polygon_mesh_processing`
* Issue(s) solved (if any): #3209, #4548

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/polygon_soup_to_polygon_mesh.h [hunk 1]
```diff
@@ -17,23 +17,99 @@
...
+// @param mesh the mesh whose faces are being put in the soup
+// @param points points of the soup of polygons
+// @param polygons each element in the vector describes a polygon using the index of the points in `points`
+//
+// \sa `CGAL::Polygon_mesh_processing::orient_polygon_soup()`
+// \sa `CGAL::Polygon_mesh_processing::is_polygon_soup_a_polygon_mesh()`
+// \sa `CGAL::Polygon_mesh_processing::polygon_soup_to_polygon_mesh()`
+//
+template<typename PolygonMesh, typename PointRange, typename PolygonRange, typename NamedParameters>
+void polygon_mesh_to_polygon_soup(const PolygonMesh& mesh,
```
Comment (MaelRL) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/polygon_soup_to_polygon_mesh.h]: Todo replace by the proper function once it is integrated (https://github.com/CGAL/cgal/pull/4418)

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h [hunk 2]
```diff
@@ -0,0 +1,1865 @@
...
+static int self_intersections_solved_by_unconstrained_smoothing = 0;
+static int self_intersections_solved_by_constrained_hole_filling = 0;
+static int self_intersections_solved_by_unconstrained_hole_filling = 0;
+#endif
+
+// -*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
+
+// @todo these could be extracted to somewhere else, it's useful in itself
+template <typename PolygonMesh, typename VPM, typename Point, typename FaceOutputIterator>
+FaceOutputIterator replace_faces_with_patch(const std::vector<typename boost::graph_traits<PolygonMesh>::vertex_descriptor>& border_vertices,
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h]: TODO to be benched with `add_faces()`

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h [hunk 3]
```diff
@@ -0,0 +1,1865 @@
...
+    for(halfedge_descriptor h : CGAL::halfedges_around_face(halfedge(f, tmesh), tmesh))
+    {
+      // Default initialization is guaranteed to be `false`. Thus, meet it once will switch
+      // the value to `true` and meeting it twice will switch back to `false`.
+      const edge_descriptor e = edge(h, tmesh);
+      is_border_of_selection[e] = !(is_border_of_selection[e]);
+    }
+  }
+
+#if 0 // Until detect_features++ is integrated
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h]: Update?

Comment (MaelRL) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/remove_self_intersections.h]: It is not yet integrated (https://github.com/CGAL/cgal/pull/4084), and I don't like the current detect_features++ enough to push for its integration.

**Comment by maxGimeno** [comment]
This PR breaks the demo : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-110/Polyhedron_Demo/TestReport_lrineau_ArchLinux-clang-Release.gz

**Comment by MaelRL** [comment]
The merge into integration is at fault here.

**Comment by maxGimeno** [comment]
Is is ? Sorry, then. My bad :/

**Comment by maxGimeno** [comment]
Succesfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-112.shtml

**Comment by maxGimeno** [comment]
warnings: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-113/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Release-64bits.gz

**Comment by sloriot** [comment]
Tested in CGAL-5.1-Ic-114

**Pull request merged**